sdk/core/System.ClientModel/src/Pipeline/ResponseBufferingPolicy.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ClientModel.Primitives;
+
+/// <summary>
+/// Pipeline policy to buffer response content or add a timeout to response content
+/// managed by the client.
+/// </summary>
+public class ResponseBufferingPolicy : PipelinePolicy
+{
+    internal static readonly ResponseBufferingPolicy Shared = new();
+
+    public ResponseBufferingPolicy()
+    {
+    }
+
+    public sealed override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+        => ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+
+    public sealed override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+        => await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        TimeSpan invocationNetworkTimeout = (TimeSpan)message.NetworkTimeout!;
```

Comment:
How do we know this isn't null? I thought it gets set by the transport so it would be null when sending the request.

Reply:
Also, if the above is accurate, please add a test that would fail with the current code.

Reply:
I believe we're setting it in pipeline.Send, so it should always get set (outside of unit tests), but I'm happy to add a proper null check to be safer here.

Reply:
Actually, I'll add a Debug.Assert to avoid branching in release builds.  LMK if you feel this doesn't sufficiently address the issue!

Reply:
Ah ok. I think a Debug.Assert would be nice to have.